src/Core/Form/IdentifiableObject/DataHandler/RootCategoryFormDataHandler.php

```diff
@@ -157,35 +185,59 @@ private function createEditRootCategoryCommand($rootCategoryId, array $data)
         $command->setLocalizedMetaDescriptions($data['meta_description']);
         $command->setLocalizedMetaKeywords($data['meta_keyword']);
         $command->setAssociatedGroupIds($data['group_association']);
-        $command->setAssociatedShopIds($data['shop_association']);
+
+        if (isset($data['shop_association'])) {
+            $command->setAssociatedShopIds($data['shop_association']);
+        }
 
         return $command;
     }
 
     /**
-     * @param int $categoryId
-     * @param UploadedFile $coverImage
-     * @param UploadedFile $thumbnailImage
+     * @param CategoryId $categoryId
+     * @param UploadedFile|null $coverImage
+     * @param UploadedFile|null $thumbnailImage
      * @param UploadedFile[] $menuThumbnailImages
      */
     private function uploadImages(
-        $categoryId,
+        CategoryId $categoryId,
         UploadedFile $coverImage = null,
         UploadedFile $thumbnailImage = null,
         array $menuThumbnailImages = []
     ) {
         if (null !== $coverImage) {
-            $this->categoryCoverUploader->upload($categoryId, $coverImage);
+            $this->categoryCoverUploader->upload($categoryId->getValue(), $coverImage);
         }
 
         if (null !== $thumbnailImage) {
-            $this->categoryThumbnailUploader->upload($categoryId, $thumbnailImage);
+            $this->categoryThumbnailUploader->upload($categoryId->getValue(), $thumbnailImage);
         }
 
         if (!empty($menuThumbnailImages)) {
             foreach ($menuThumbnailImages as $menuThumbnail) {
-                $this->categoryMenuThumbnailUploader->upload($categoryId, $menuThumbnail);
+                $this->categoryMenuThumbnailUploader->upload($categoryId->getValue(), $menuThumbnail);
+            }
+        }
+    }
+
+    /**
+     * @param int $categoryId
+     *
+     * @return array<int, int>
+     */
+    private function getAvailableKeys(int $categoryId): array
+    {
+        $files = scandir(_PS_CAT_IMG_DIR_, SCANDIR_SORT_NONE);
```

Comment:
same here, use DI and finder, maybe this code can be refactored, looks like it has been copy pasted

Reply:
Thanks! Refactored the code